numpy/core/src/umath/ufunc_object.c

```diff
@@ -551,6 +551,201 @@ ufunc_get_name_cstr(PyUFuncObject *ufunc) {
+               (cmp = PyObject_RichCompareBool(key, **kwname,
+                                               Py_EQ)) == 0) {
+            kwname++;
+        }
+        if (cmp < 0) {
+            return -1;
+        }
+    }
+    return kwname - kwnames;
+}
+
+/*
+ * Parse keyword arguments, matching against kwnames
+ *
+ * Arguments beyond kwnames (the va_list) should contain converters and outputs
+ * for each keyword name (where an output can be NULL to indicate the particular
+ * keyword should be ignored).
+ *
+ * Returns 0 on success, -1 on failure (with an error set and any already
+ * parsed keyword values DECREF'd by calling converters using obj=NULL).
+ *
+ * TODO: ufunc is only used for the name in error messages; passing on the
+ *       name instead might be an option.
+ *
+ * TODO: instead of having this function ignore of keywords for which the
+ *       corresponding output is NULL, the calling routine should prepare the
+ *       correct list.
+ */
+static int
+parse_ufunc_keywords(PyUFuncObject *ufunc, PyObject *kwds, PyObject ***kwnames, ...)
+{
+    va_list va;
+    PyObject *key, *value;
+    Py_ssize_t pos = 0;
+    typedef int converter(PyObject *, void *);
+
+    while (PyDict_Next(kwds, &pos, &key, &value)) {
+        int i;
+        converter *convert;
+        void *output = NULL;
+        npy_intp index = locate_key(kwnames, key);
+        if (index < 0) {
+            return -1;
+        }
+        if (kwnames[index]) {
+            va_start(va, kwnames);
+            for (i = 0; i <= index; i++) {
+                convert = va_arg(va, converter *);
+                output = va_arg(va, void *);
+            }
+            va_end(va);
+        }
+        if (output) {
+            if (!convert(value, output)) {
+                goto fail;
+            }
+        }
+        else {
+#if PY_VERSION_HEX >= 0x03000000
+            PyErr_Format(PyExc_TypeError,
+                         "'%S' is an invalid keyword to ufunc '%s'",
+                         key, ufunc_get_name_cstr(ufunc));
+#else
+            char *str = PyString_AsString(key);
+            if (str == NULL) {
+                PyErr_Clear();
+                PyErr_SetString(PyExc_TypeError, "invalid keyword argument");
+            }
+            else {
+                PyErr_Format(PyExc_TypeError,
+                             "'%s' is an invalid keyword to ufunc '%s'",
+                             str, ufunc_get_name_cstr(ufunc));
+            }
+#endif
+            goto fail;
+        }
+    }
+    return 0;
+
+fail:
+    /*
+     * If we already treated some keywords, repeat the scan of the dict
+     * for everything done previously and call their converters with value=NULL
+     * to indicate they should XDECREF outputs as needed (similar to what is
+     * done in PyArg_ParseTupleAndKeywords).
+     */
+    if (pos != 0) {
+        Py_ssize_t repeat = 0;
+        while (PyDict_Next(kwds, &repeat, &key, &value) && repeat != pos) {
+            int i;
```

Comment:
Obviously, some repetition here with the above. Not quite sure how to solve it.

Reply:
Could encapsulate the loop::
```
for (i = 0; i <= index; i++) {
    convert = va_arg(va, converter *);
    output = va_arg(va, void *);
}
```
But arguably not worth it.